src/raps/execute.ts

```diff
@@ -0,0 +1,166 @@
+import {
+  ActionProps,
+  Rap,
+  RapAction,
+  RapActionResponse,
+  RapActionResult,
+  RapActionTypes,
+  RapSwapActionParameters,
+  RapTypes,
+} from './references';
+import { createUnlockAndCrosschainSwapRap } from './unlockAndCrosschainSwap';
+import { createUnlockAndSwapRap } from './unlockAndSwap';
+import { GasFeeParamsBySpeed, LegacyGasFeeParamsBySpeed, LegacySelectedGasFee, SelectedGasFee } from '@/entities';
+
+export function createSwapRapByType<T extends RapTypes>(type: T, swapParameters: RapSwapActionParameters<T>) {
+  switch (type) {
+    case 'crosschainSwap':
+      return createUnlockAndCrosschainSwapRap(swapParameters as RapSwapActionParameters<'crosschainSwap'>);
+    case 'swap':
+      return createUnlockAndSwapRap(swapParameters as RapSwapActionParameters<'swap'>);
+    default:
+      return { actions: [] };
+  }
+}
+
+function typeAction<T extends RapActionTypes>(type: T, props: ActionProps<T>) {
+  switch (type) {
+    case 'unlock':
+      return () => unlock(props as ActionProps<'unlock'>);
+    case 'swap':
+      return () => swap(props as ActionProps<'swap'>);
+    case 'crosschainSwap':
+      return () => crosschainSwap(props as ActionProps<'crosschainSwap'>);
+    default:
+      // eslint-disable-next-line react/display-name
+      return () => null;
+  }
+}
+
+export async function executeAction<T extends RapActionTypes>({
+  action,
+  wallet,
+  rap,
+  index,
+  baseNonce,
+  rapName,
+  flashbots,
+  selectedGasFee,
+  gasFeeParamsBySpeed,
+}: {
+  action: RapAction<T>;
+  wallet: Signer;
+  rap: Rap;
+  index: number;
+  baseNonce?: number;
+  rapName: string;
+  flashbots?: boolean;
+  selectedGasFee: SelectedGasFee | LegacySelectedGasFee;
+  gasFeeParamsBySpeed: GasFeeParamsBySpeed | LegacyGasFeeParamsBySpeed;
+}): Promise<RapActionResponse> {
+  const { type, parameters } = action;
+  try {
+    const actionProps = {
+      wallet,
+      currentRap: rap,
+      index,
+      parameters: { ...parameters, flashbots },
+      baseNonce,
+      selectedGasFee,
+      gasFeeParamsBySpeed,
+    };
+    const { nonce, hash } = (await typeAction<T>(type, actionProps)()) as RapActionResult;
+    return { baseNonce: nonce, errorMessage: null, hash };
+  } catch (error) {
+    logger.error(new RainbowError(`rap: ${rapName} - error execute action`), {
+      message: (error as Error)?.message,
+    });
+    if (index === 0) {
+      return { baseNonce: null, errorMessage: String(error) };
+    }
+    return { baseNonce: null, errorMessage: null };
+  }
+}
+
+function getRapFullName<T extends RapActionTypes>(actions: RapAction<T>[]) {
+  const actionTypes = actions.map(action => action.type);
+  return actionTypes.join(' + ');
+}
+
+const delay = (ms: number) => new Promise(res => setTimeout(res, ms));
+
+const waitForNodeAck = async (hash: string, provider: Signer['provider']): Promise<void> => {
+  return new Promise(async resolve => {
+    const tx = await provider?.getTransaction(hash);
```

Comment:
why not `provider.waitForTransaction`?
https://docs.ethers.org/v5/api/providers/provider/#Provider-waitForTransaction

Reply:
We could use that and wait for 1 confirmation, however for the sake of keeping code parity between app <> bx I didn't change any of the raps (functionally).

See: https://github.com/rainbow-me/browser-extension/blob/09b452e60561671d6ae25e9dc8915cde31683920/src/core/raps/execute.ts#L109